src/twin/substrate.rs

```diff
@@ -64,12 +64,50 @@ impl StorageEntry for TwinAccountID {
 }
 
 type Client = subxt::Client<DefaultConfig>;
+
+#[derive(Debug, Clone)]
+struct ReconnectingClient {
+    client: Arc<RwLock<Client>>,
+    url: String,
+}
+
+impl ReconnectingClient {
+    async fn get_new(url: &str) -> Result<Client> {
+        Ok(ClientBuilder::new().set_url(url).build().await?)
+    }
+
+    pub async fn new<S: Into<String>>(url: S) -> Result<Self> {
+        let url = url.into();
+
+        let client = Self::get_new(&url).await?;
+        Ok(Self {
+            client: Arc::new(RwLock::new(client)),
+            url,
+        })
+    }
+
+    async fn is_connected(&self) -> bool {
+        self.client.read().await.rpc().client.is_connected()
+    }
+
+    pub async fn get(&self) -> Result<Client> {
+        if !self.is_connected().await {
+            log::debug!("client is disconnected, rebuilding a new one");
+
+            let mut client = self.client.write().await;
+            *client = Self::get_new(&self.url).await?;
```

Comment:
This is still a possible race here because the "check" if the connection is valid and the "update" of the connection is not atomic. Let me explain
- 1) You acquire a read lock 
- You check the connection
- You release the lock.
- 2) You found that the connection is down, so you decide to take a write lock 
- You renew the connection
 (Now imagine there is another thread is blocked at step "1" above. Because it's also trying to get a read lock, it will succeed because read lock is shared)
- the second thread is now waiting on the write lock which is acquired by the first thread and is actively creating a new connection
- The write lock is release so the second thread is now free to take the lock and then try to renew the connection
- You end up with 2 connections.

The easiest solution to this is to do the following
- Acquire a lock 
- check connection
- renew if needed
- release

(Since there is no need for read/write lock in this case you can use Mutex instead of RwLock

Reply:
I don't think [is_connected](https://github.com/paritytech/jsonrpsee/blob/3c416ac56fb2bb678a051afaf9940af3c3daf45d/core/src/client/async_client/mod.rs#L224) does any extra calls